crates/oxc_linter/src/rules/eslint/curly.rs

```diff
@@ -0,0 +1,1904 @@
+            }
+
+            if let Some(stripped) = fixed.strip_suffix(|c: char| c.is_whitespace() || c == '}') {
+                fixed = stripped.to_string();
+            } else if fixed.ends_with('}') {
+                fixed.pop();
+            }
+
+            if needs_preceding_space {
+                fixed.insert(0, ' ');
+            }
+            fixer.replace(body.span(), fixed)
+        }
+    });
+}
+
+fn is_collapsed_one_liner(node: &Statement, ctx: &LintContext) -> bool {
+    let node = get_node_by_statement(node, ctx);
+    let span = node.span();
+    let node_string = ctx.source_range(span);
+
+    let trimmed_len =
+        u32::try_from(node_string.trim_end_matches(|c: char| c.is_whitespace() || c == ';').len())
+            .expect("Failed to convert to u32");
+
+    let before_node_span = get_token_before(node, ctx).map_or_else(
+        || {
+            ctx.nodes()
+                .parent_node(node.id())
+                .filter(|parent| parent.span().start < span.start)
+                .map_or(Span::new(0, 0), |parent| {
+                    Span::new(parent.span().start, parent.span().start)
+                })
+        },
+        oxc_span::GetSpan::span,
+    );
+
+    let text = ctx.source_range(Span::new(
+        before_node_span.end + 1,
+        span.end
+            - (u32::try_from(node_string.len()).expect("Failed to convert to u32") - trimmed_len),
+    ));
+
+    !text.contains('\n')
+}
+
+fn is_one_liner(node: &Statement, ctx: &LintContext) -> bool {
+    if matches!(node, Statement::EmptyStatement(_)) {
+        return true;
+    }
+
+    let source = ctx.source_range(node.span());
+    let trimmed = source.trim_end_matches(|c: char| c.is_whitespace() || c == ';');
+    !trimmed.contains('\n')
+}
+
+fn get_token_before<'a>(node: &AstNode, ctx: &'a LintContext) -> Option<&'a AstNode<'a>> {
+    let span_start = node.span().start;
+    ctx.nodes().iter().filter(|n| n.span().end < span_start).max_by_key(|n| n.span().end)
+}
+
+pub fn are_braces_necessary(root_node: &AstNode, node: &Statement, ctx: &LintContext) -> bool {
+    let Statement::BlockStatement(block) = node else {
+        return false;
+    };
+
+    let Some(first_body_statement) = block.body.first() else {
+        return false;
+    };
+
+    is_lexical_declaration(first_body_statement)
+        || (has_unsafe_if(first_body_statement)
+            && is_followed_by_else_keyword(root_node, node, ctx))
+}
+
+fn is_lexical_declaration(node: &Statement) -> bool {
+    match node {
+        Statement::VariableDeclaration(decl) => {
+            matches!(decl.kind, VariableDeclarationKind::Const | VariableDeclarationKind::Let)
+        }
+        Statement::FunctionDeclaration(_) | Statement::ClassDeclaration(_) => true,
+        _ => false,
+    }
+}
+
+fn is_followed_by_else_keyword(root_node: &AstNode, node: &Statement, ctx: &LintContext) -> bool {
+    let start = node.span().end + 1;
+    let end = root_node.span().end;
```

Comment:
I don't think the root node is necessary here? You can get the last byte offset with `ctx.source_text().len()`

Reply:
fixed: https://github.com/oxc-project/oxc/commit/d760bc434f99e25970e59bce3e0c7e933f327f12